File: L10-PrimeAndCompositeNumbers/CountFactors.cpp
// Date: Nov 29, 2021 -> Redo: Dec 7, 2021
// CountFactors

// Given N = 24, the function should return 8, 
// because 24 has 8 factors, namely 1, 2, 3, 4, 6, 8, 12, 24. 

// N = [1..2,147,483,647]

// time complexity: O(sqrt(n))
// space complexity: O(1)
// result: https://app.codility.com/demo/results/training5PTD6J-8VZ/
int solution(int N) {
	int ans = 0;
    for (long i = 1; i * i <= N; i++) {
        if (i * i == N)
            ans += 1;
        else if (N % i == 0)
            ans += 2;
    }
    return ans;
}

// last result: 92%
// failed test case:
// For example, for the input 2147483647 the solution exceeded the time limit.

// reference
// https://github.com/Behrouz-m/Codility/blob/master/10-PrimeAndCompositeNumbers/CountFactors.cpp

File: L10-PrimeAndCompositeNumbers/CountFactors.swift
import Foundation
// result: https://app.codility.com/demo/results/trainingKMYSD9-2VX/
public class CountFactors {
    public func solution(_ N : Int) -> Int {
        var count = 0
        var first = 1
        while first * first <= N {
            if N % first == 0 {
                if first * first == N {
                    count += 1
                } else {
                    count += 2
                }
            }
            first += 1
        }
        return count
    }
}

// given N = 24, the function should return 8, because 24 has 8 factors, namely 1, 2, 3, 4, 6, 8, 12, 24
